### PR TITLE
Update schema for snapshots and tarballs

### PR DESCRIPTION
**Pull request by elric1**
Our new schema separates setting the URLs from deciding when and
how to use them.

### CODE REVIEW

**Comment by elric1** [comment]
Please note that the helm lint failure is solved in my prior pull request #393 which I expect we will review and merge before this one. I will need to rebase a little in any case, but there's enough here to start a discussion of the approach.

File: charts/tezos/values.yaml [hunk 1]
```diff
@@ -93,6 +93,8 @@ accounts: {}
 ##      automatically for you.
 ## - "node_selector": Specify a kubernetes node selector in 'key: value' format
 ##     for your tezos nodes.
+## - "snapshot": enum of "snapshot", "tarball", or "none" defaulting to none.
```
Comment (harryttd) [inline charts/tezos/values.yaml:96]: Tezos snapshots are named so aptly well, snapshots. So it has made naming and referencing things harder in regards to tarballs which are also snapshots, but not "tezos snapshots".

Can we use the word 'standard' in place of 'snapshot' as a value so we don't have this sort of self-referencing thing going on? I think the word gives us the right category term to differentiate between regular snapshots and tarballs.

**Review by harryttd — CHANGES_REQUESTED** [review]
great pr :)

**Comment by orcutt989** [comment]
Great PR thanks @elric1

**Comment by elric1** [comment]
Interesting, the tests operate not against the branch itself, but rather against the branch as merged against HEAD. I've rebased and force pushed because the tests needed the new default image to be v13-release.

File: charts/tezos/values.yaml (hunk 1 above)
Comment (elric1) [reply charts/tezos/values.yaml:128]: I put "none" in specifically rather than just let it be only null. That's how a node can sync from scratch even if a global value is set to use snapshots.

File: charts/tezos/templates/_helpers.tpl [hunk 2]
```diff
@@ -68,12 +68,10 @@
   Returns a string "true" or empty string which is falsey.
 */}}
 {{- define "tezos.shouldDownloadSnapshot" -}}
-  {{- if or (.Values.full_snapshot_url) (.Values.rolling_snapshot_url) (.Values.rolling_tarball_url) (.Values.archive_tarball_url) }}
-    {{- if and (.Values.rolling_tarball_url) (.Values.rolling_snapshot_url) }}
-      {{- fail "Either only a snapshot url or tarball url may be specified per Tezos node history mode" }}
-    {{- else }}
-      {{- "true" }}
-    {{- end }}
+  {{- if or (hasKey $.node_vals "snapshot")
```
[the remark this answers is not in the document]
Comment (elric1) [reply charts/tezos/templates/_helpers.tpl]: This is commit `8b93d59d88`. I'll squash it in later because there's not need to immortalise the intermediate state.

File: utils/snapshot-downloader.sh [hunk 3]
```diff
@@ -19,50 +39,38 @@ fi
...
+  SNAPSHOT_TYPE=$(echo $NODE_GLOBALS | jq -r .snapshot)
 fi
 
-if [ -n "$snapshot_url" ] && [ -n "$tarball_url" ]; then
-  echo "ERROR: Either only a snapshot or tarball url may be specified per Tezos node history mode."
+if [ "$SNAPSHOT_TYPE" != snapshot -a "$SNAPSHOT_TYPE" != tarball ]; then
+  echo -n "ERROR: Snapshot type (\"$SNAPSHOT_TYPE\") invalid:"
+  echo 'must be "snapshot" or "tarball"'
+  exit 1
 fi
```
[the remark this answers is not in the document]
Comment (elric1) [reply utils/snapshot-downloader.sh]: d1843a9 should take care of this. The code has changed a bit since the initial comment, but you can see the intent.

**Review by nicolasochem — COMMENTED** [review]
From an usability perspective this is good, I like that I can just write a very small values.yaml for deploying an ithacanet baker, and the snapshots/tarball links are already set for you. No need to put them in your values file.

If I am using a custom/unknown chain I can still add a map under `snapshots:` with my chain name as key, and specify snapshots at will. It's a bit harder to use than today, but on the other hand I am not sure anyone is doing this.

File: charts/tezos/values.yaml (hunk 1 above)
Comment (elric1) [reply charts/tezos/values.yaml]: fae64dc renames snapshot to standard. We could consider other names, though, such as "tezos" or something like that. I don't have much of an opinion, though, I think that standard should be good enough.

Comment (harryttd) [reply charts/tezos/values.yaml]: `snapshot: octez/tezos` can sound simply like "use a snapshot of the Tezos blockchain" as opposed to really differentiating between the type of snapshot, `standard` vs `tarball`. Tarball is also a Tezos snapshot of the blockchain just formatted differently. So i think leaving as `standard` would be fine IMO

Comment (harryttd) [reply charts/tezos/values.yaml]: > I put "none" in specifically rather than just let it be only null. That's how a node can sync from scratch even if a global value is set to use snapshots.

Just making sure i understand. Wouldn't this work:
Instead of making jq return `null` if no snapshot is set, return either `empty` or empty string. And just check that `SNAPSHOT_TYPE` is `empty` / doesn't has a value set. If yes, then use the global value for snapshot.

Otherwise if values.y `snapshot: null`, then `null` is a valid option, `SNAPSHOT_TYPE == null`, and the following code would use `null` instead of `none`. 
```shell
case "$SNAPSHOT_TYPE" in
standard)	if [ "$HISTORY_MODE" = archive ]; then
		    bail "Standard snapshots can't do archive mode, " \
		         "use tarball instead."
		fi
		standard "$SNAPSHOT_URL";;
tarball)	tarball "$SNAPSHOT_URL";;
null)		;;
*)
```

If i'm looking at this correctly, wouldn't this be simpler than introducing a new value for the enum instead of relying on an existing yaml primitive?

Comment (elric1) [reply charts/tezos/values.yaml]: null and none are different. null means use the value from the higher level, none means do not use a snapshot at all. The reason that null means that it is that it typically indicates that at this level, no one made a decision. I.e.:
node_globals:
  snapshot: tarball
nodes:
  cape-town:
    snapshot: none
    instances:
      - {}
will start cape-town-0 with no snapshot. If I set it to null, then it would use a tarball.

File: utils/snapshot-downloader.sh [hunk 4]
```diff
@@ -19,49 +39,66 @@ fi
...
-  *)        echo "Invalid node history mode: '$my_nodes_history_mode'"
+  *)        echo "Invalid node history mode: '$HISTORY_MODE'"
             exit 1;;
 esac
 
-if [ -z "$snapshot_url" ] && [ -z "$tarball_url" ]; then
-  echo "ERROR: No snapshot or tarball url specified."
-  exit 1
+if [ -n "$snapshot_url" ] && [ -n "$tarball_url" ]; then
+  bail "Only a snapshot or tarball url may be specified per history mode."
```
Comment (harryttd) [inline utils/snapshot-downloader.sh:73]: Perhaps should have additional text? Saying something along the lines of this error applies only when using the deprecated fields.

File: utils/snapshot-downloader.sh [hunk 5]
```diff
@@ -19,49 +39,66 @@ fi
...
   archive)  tarball_url="$ARCHIVE_TARBALL_URL";;
-
-  *)        echo "Invalid node history mode: '$my_nodes_history_mode'"
+  *)        echo "Invalid node history mode: '$HISTORY_MODE'"
             exit 1;;
 esac
 
-if [ -z "$snapshot_url" ] && [ -z "$tarball_url" ]; then
-  echo "ERROR: No snapshot or tarball url specified."
-  exit 1
```
Comment (harryttd) [inline utils/snapshot-downloader.sh:48]: I think we still need a check like this after all of our logic checking where the url comes from. Saying, that if a snapshot type was specified for X history mode, then a url needs to exist for it. As far as i can tell, there is no check for this.

**Review by harryttd — CHANGES_REQUESTED** [review]
Thank you for this great PR :) 

Please see my comments.

One additional thought is, perhaps organize the deprecated tests into a subfolder? Perhaps /test/charts/deprecated or something?

File: charts/tezos/templates/configs.yaml [hunk 6]
```diff
@@ -16,6 +16,11 @@ data:
   ROLLING_SNAPSHOT_URL: "{{ .Values.rolling_snapshot_url }}"
   ROLLING_TARBALL_URL: "{{ .Values.rolling_tarball_url }}"
   ARCHIVE_TARBALL_URL: "{{ .Values.archive_tarball_url }}"
+  SNAPSHOTS: |
+{{ get .Values.snapshots .Values.node_config_network.chain_name
+	| default (get .Values.snapshots "default")
```
[the remark this answers is not in the document]
Comment (elric1) [reply charts/tezos/templates/configs.yaml]: Oh, I put that in just in case you have an odd chain name like:
```
node_config_network:
  chain_name: https://teztnets.xyz/dailynet-2022-02-08
```
default matches any name. It's a decent thing to put in your own values.yaml file because you may not be setting up a number of different chains and therefore may not need to match chain names.
Whitespace is probably a tab, I use those a lot. We usually take them out of the YAML parts because they break things, but they make no difference inside `{{` and `}}`.
I expanded the tabs in c7b98ab which will be squashed before the release.

File: utils/snapshot-downloader.sh (hunk 5 above)
Comment (elric1) [reply utils/snapshot-downloader.sh:48]: Heh, well, curl will output 'No URL specified', but, perhaps a better error might be interesting. I've added this enhanced error reporting in bc04075.

File: charts/tezos/values.yaml [hunk 7]
```diff
@@ -237,17 +240,26 @@ signers: {}
...
-full_snapshot_url: https://mainnet.xtz-shots.io/full
-rolling_snapshot_url: https://mainnet.xtz-shots.io/rolling
+##
+## Where to find snapshots of various forms.  We call it "standard" if
+## it's a standard Tezos snapshot.  A tarball is an LZ4-compressed tarball
+## which is unpacked without being checked.  It is therefore faster and
+## less safe.  This data structure merely defines the URLs, not whether
+## to actually use a standard or tarball.
+##
+## NOTE: there is no "standard" for archive because it doesn't make sense.
```
[the remark this answers is not in the document]
Comment (elric1) [reply charts/tezos/values.yaml]: Did this in d4fcd24.

File: charts/tezos/values.yaml [hunk 8]
```diff
@@ -237,17 +240,26 @@ signers: {}
...
+
+snapshots:
+  mainnet:
+    rolling:
+      standard: https://mainnet.xtz-shots.io/rolling
+      tarball:  https://mainnet.xtz-shots.io/rolling-tarball
+    full:
+      standard: https://mainnet.xtz-shots.io/full
+      tarball:  https://mainnet.xtz-shots.io/full-tarball
```
[the remark this answers is not in the document]
Comment (elric1) [reply charts/tezos/values.yaml:259]: Or maybe add full snapshots to xtz-shots?

[the remark this answers is not in the document]
Comment (harryttd) [reply charts/tezos/values.yaml:259]: yes :) But that takes some work on the snapshot engine side and would need to be coordinated.

File: charts/tezos/values.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (harryttd) [reply charts/tezos/values.yaml]: https://github.com/oxheadalpha/tezos-k8s/blob/d4fcd24c489741446e1e3370a4f9e1eb880bf0ef/charts/tezos/values.yaml#L250
"it aren't" -> "it isn't"

[the remark this answers is not in the document]
Comment (elric1) [reply charts/tezos/values.yaml]: 5b1b27d

File: charts/tezos/values.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (elric1) [reply charts/tezos/values.yaml]: e3bb84b

File: charts/tezos/templates/_helpers.tpl [hunk 9]
```diff
@@ -78,6 +82,20 @@
...
     {{- end }}
+  {{- end }}
+{{- end }}
+
+{{- define "tezos.shouldDownloadSnapshot" -}}
+  {{- $snapshotType :=
+               or (get $.node_vals "snapshot")
+                  (get (get .Values "node_globals" | default dict) "snapshot")
+                  "none" }}
+  {{- if get (dict "standard" 1 "tarball" 1 "none" 1) $snapshotType | not }}
```
[the remark this answers is not in the document]
Comment (elric1) [reply charts/tezos/templates/_helpers.tpl]: ecc7b15a7098

File: utils/snapshot-downloader.sh [hunk 10]
```diff
@@ -19,49 +52,65 @@ fi
...
-  curl -LfsS -o "$snapshot_file" "$snapshot_url"
-elif [ -n "$tarball_url" ]; then
-  echo "Downloading and extracting tarball from $tarball_url"
-  curl -LfsS "$tarball_url" | lz4 -d | tar -x -C "$data_dir"
-fi
+case "$SNAPSHOT_TYPE" in
+standard)	standard "$SNAPSHOT_URL";;
+tarball)	tarball "$SNAPSHOT_URL";;
+none)		;;
+*)		usage "$SNAPSHOT_TYPE must be 'standard', 'tarball',"
```
Comment (harryttd) [inline utils/snapshot-downloader.sh:111]: 1.
```shell
│ + CMD=snapshot-downloader                                                                                                 │
│ + shift                                                                                                                   │
│ + exec /snapshot-downloader.sh                                                                                            │
│ Did not find a pre-existing blockchain.                                                                                   │
│ My nodes history mode: 'rolling'                                                                                          │
│ /snapshot-downloader.sh: line 111: usage: not found                                                                       │
│ Stream closed EOF for x/rolling-node-0 (snapshot-downloader)                                                              │
│                                                                                                                           │
│
```
2.
```shell
Did not find a pre-existing blockchain.                                                                                   │
│ + CMD=snapshot-downloader                                                                                                 │
│ + shift                                                                                                                   │
│ + exec /snapshot-downloader.sh                                                                                            │
│ My nodes history mode: 'rolling'                                                                                          │
│ s                                                                                                                         │
│ /snapshot-downloader.sh: line 114: or 'none'.: not found                                                                  │
│ Stream closed EOF for x/rolling-node-0 (snapshot-downloader)
```
Screen Shot 2022-05-31 at 1 13 00 PM

**Review by harryttd — CHANGES_REQUESTED** [review]

File: charts/tezos/templates/configs.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (harryttd) [reply charts/tezos/templates/configs.yaml]: FYI this works in Helm:
```yaml

snapshots:
  https://teztnets.xyz/mondaynet:
    rolling:
      tarball: https://....

node_config_network:
  chain_name: "https://teztnets.xyz/mondaynet"
```

Also do you mind clarifying: 
> It's a decent thing to put in your own values.yaml file because you may not be setting up a number of different chains and therefore may not need to match chain names.

I'm not sure i follow what this case means. Do you by any chance mean "you may ~not~ be setting up..."? What would a case look like where i would use default?

File: utils/snapshot-downloader.sh (hunk 10 above)
Comment (elric1) [reply utils/snapshot-downloader.sh]: Hmmm, never hit that code path.

File: charts/tezos/templates/configs.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (elric1) [reply charts/tezos/templates/configs.yaml]: I suppose that the use-case for default isn't terribly strong. I am imaging that if you want to override the snapshot location in your own values file, then it's a little redundant to make you specify which chain that you are using---you've already done it. So, you can just use "default" rather than duplicating whatever chain name that you've used. It's not much of an optimisation, but, I put it in.

File: utils/snapshot-downloader.sh (hunk 10 above)
Comment (harryttd) [reply utils/snapshot-downloader.sh]: i made sure to test

**Review by harryttd — APPROVED** [review]

**Review by nicolasochem — COMMENTED** [review]
I tried this and it worked:

```
node_config_network:
  chain_name: "jakartanet"

tezos_k8s_images:
  utils: localhost/tezos-k8s-utils:dev
    #rolling_snapshot_url: https://jakartanet.xtz-shots.io/rolling
snapshots:
  jakartanet:
    rolling:
      standard: https://jakartanet.xtz-shots.io/rolling

images:
  octez: tezos/tezos:v13-release

nodes:
  rolling-node:
    runs:
      - octez_node
    instances:
      - is_bootstrap_node: true
        config:
          shell:
            history_mode: rolling
```

But the old method didn't work however:

```
node_config_network:
  chain_name: "jakartanet"

tezos_k8s_images:
  utils: localhost/tezos-k8s-utils:dev
rolling_snapshot_url: https://jakartanet.xtz-shots.io/rolling

images:
  octez: tezos/tezos:v13-release

nodes:
  rolling-node:
    runs:
      - octez_node
    instances:
      - is_bootstrap_node: true
        config:
          shell:
            history_mode: rolling
```

I thought we were preserving backward compatibility for now. Are we?

Also do we want ithacanet/jakartanet URLs from xtz-shots hardcoded too?

**Comment by elric1** [comment]
Hmmm, well, I did try to preserve backwards compat, but this is an edge case that I didn't test. What's going on here is that my logic is: if you do not define the snapshot type in the node config or node_globals, then I fall back to the old mechanism. _BUT_ what appears to happen here (which I didn't test for, obviously) is that even though you define a "rolling-node" statefulset, it doesn't override all of the defaults for it. That means that even though you didn't specify it in your values.yaml:
```
nodes:
  rolling-node:
    snapshot: standard
```
is specified.
I think that maybe I will have to change the sense of the conditional and make it s.t. if you define `rolling_snapshot_url`, etc, they will override the new mechanism. I'm not fond of that, because I'd rather see the new mechanism take priority over the old, but I don't see a way to do that whilst retaining backwards compat.

**Review by nicolasochem — CHANGES_REQUESTED** [review]
Thank you for explaining the issue I had with the old model. I think you are right and it's better to have the new model take precedence.

I think that if we provide default xtz-shots urls for mainnet, ithacanet and jakartanet, then most users will have to muck with these params or care about their existence. We should not spend extra effort just for bw compat. So let's just hardcode ithacanet and jakartanet and merge.

**Comment by nicolasochem** [comment]
Another thought: this change restricts usage of snapshots to "known" chains which are hardcoded in octez (mainnet, jakartanet).

If someone wants to use tezos-k8s to create a totally private chain and maintains snapshots for this chain, then there is no longer any any way to pass this snapshot as parameter and bootstrap a private chain node from snapshot/tarball.

I don't think anyone is doing it, but did we consider this? Is it a problem?

**Comment by harryttd** [comment]
@nicolasochem see Roland and my convo here: https://github.com/oxheadalpha/tezos-k8s/pull/394#discussion_r883064047

Does that answer your question?

**Comment by nicolasochem** [comment]
@harryttd oh yes, it does. I guess it needs to be documented?

**Comment by elric1** [comment]
I want to make another proposal for the backwards compat mechanism.
As I mentioned above, I don't want `rolling_snapshot_url`, et al. at the top level override people's specific setting of:
```
nodes:
  rolling:
    snapshot: tarball
```
But, I do want to provide a default for the chart with just "rolling-node" defined.
Right now, `nodes-><name>->snapshot` is an enumerated type with "standard", "tarball", or "none" as possible values. I am going to propose that for backwards compat, I add a value of "compat" which has the semantics of: "standard" unless `rolling_snapshot_url` is defined in which case that value takes precedence. In this way, we will use the legacy values if `nodes-><name>->snapshot` isn't defined or if it is set to "compat". The difference between undefined and "compat" is that undefined means no snapshots unless the legacy url values are defined whereas "compat" means standard if they aren't defined.

**Review by nicolasochem — APPROVED** [review]
@elric1 the new proposal for handling backward compatibility sounds good, however I think we can merge this as is. The project would be in better shape with this merged, since users can now deploy testnet setups without specifying any URL at all. Then, why add code to handle the old-style parameters? Presumably, if a user has these parameters set, they would be set to the same values than the new defaults?

**Comment by harryttd** [comment]
> @harryttd oh yes, it does. I guess it needs to be documented?

Let's add documentation

**Comment by nicolasochem** [comment]
@elric1 let's wrap this up?
We need to add ghostnet support.

Additionally, this will enable a one-liner helm in the new [doc website](https://news.tezoscommons.org/double-baking-and-remote-signers-639bf8ef7e65): a single line of helm to install a ghostnet rolling node in a cluster (no values.yaml needed).

**Comment by nicolasochem** [comment]
replaced by #525
